pkg/gardenlet/controller/shoot/shoot/reconciler_delete.go

```diff
@@ -229,12 +232,17 @@ func (r *Reconciler) runDeleteShootFlow(ctx context.Context, o *operation.Operat
 		// controller-manager to be updateable due to provider config injection.
 		deployControlPlane = g.Add(flow.Task{
 			Name:         "Deploying Shoot control plane",
-			Fn:           flow.TaskFn(botanist.DeployControlPlane).RetryUntilTimeout(defaultInterval, defaultTimeout).DoIf(cleanupShootResources && controlPlaneDeploymentNeeded),
+			Fn:           flow.TaskFn(botanist.DeployControlPlane).RetryUntilTimeout(defaultInterval, defaultTimeout).DoIf(cleanupShootResources && controlPlaneDeploymentNeeded).SkipIf(o.Shoot.IsWorkerless),
 			Dependencies: flow.NewTaskIDs(initializeSecretsManagement, deployCloudProviderSecret, ensureShootClusterIdentity),
 		})
 		waitUntilControlPlaneReady = g.Add(flow.Task{
-			Name:         "Waiting until Shoot control plane has been reconciled",
-			Fn:           flow.TaskFn(botanist.Shoot.Components.Extensions.ControlPlane.Wait).DoIf(cleanupShootResources && controlPlaneDeploymentNeeded),
+			Name: "Waiting until Shoot control plane has been reconciled",
+			Fn: flow.TaskFn(func(ctx context.Context) error {
+				if o.Shoot.IsWorkerless {
+					return nil
+				}
```

Comment:
Why don't you use `SkipIf/DoIf`? Same for other occurrences

Reply:
`botanist.Shoot.Components.Extensions.Infrastructure` won't be populated in case of workerless, so we can't call Get from it.

Reply:
We found https://github.com/gardener/gardener/pull/7882/#discussion_r1190988492 to get more context (while your response is still not quite enlightening 😅). Anyways, please use `SkipIf`. The function is not even executed when the skip condition is true, hence `botanist.Shoot.Components.Extensions.ControlPlane.Wait` will never be tried to be called. Same for other occurrences.

Reply:
SkipIf can't be used because `task` is added to the graph independent of what is present in `SkipIf` or `DoIf`,
https://github.com/gardener/gardener/blob/5abed1bfedd6f2188e5dcec9af161d1b5c286236/pkg/utils/flow/graph.go#L67-L80.

So while adding the taskFn to graph it will panic, if `SkipIf` is used cause `ControlPlane` will be nil for workerless shoot and fetching `Wait` from nil will cause the panic.

Reply:
Yes we are not executing the task, but [here](https://github.com/gardener/gardener/blob/5abed1bfedd6f2188e5dcec9af161d1b5c286236/pkg/utils/flow/graph.go#L82-L101) before executing/running when we compile the graph, we are fetching the function of the taskFn and doing: https://github.com/gardener/gardener/blob/5abed1bfedd6f2188e5dcec9af161d1b5c286236/pkg/utils/flow/graph.go#L93

Here since `ControlPlane` is nil `ControlPlane.Wait` will cause panic and similar for other occurrences.